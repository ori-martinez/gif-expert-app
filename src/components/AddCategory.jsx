/* Hooks */
import { useState } from 'react';
/* Tipos de Datos de las Propiedades del Componente */
import PropTypes from 'prop-types';

// ESTILOS
const formStyle = "flex justify-center";
const inputStyle = "p-1 pl-2 my-4 w-2/5 text-lg text-stone-900 border border-stone-900 rounded-lg placeholder:text-stone-400 focus:outline-none focus:ring focus:ring-violet-300";

// COMPONENTE
/* Input de Búsqueda de Categorías */
export const AddCategory = ({ handleCategory }) => {
    // CONSTANTES
    /* Estado Inicial del Input */
    const [ inputValue, setInputValue ] = useState('');

    // FUNCIONES
    /* Manejador de Cambio del Input */
    const handleChange = ({ target }) => {
        setInputValue(target.value);
    }
    /* Manejador de Envío del Valor Buscado */
    const handleSubmit = ( event ) => {
        event.preventDefault();

        if (inputValue.trim().length < 1) return;

        handleCategory(inputValue.trim());
        setInputValue('');
    }

    // RETORNO
    return (
        <form 
            className={ formStyle }
            onSubmit={ handleSubmit }
            aria-label="form"
        >
            <input
                type="text"
                placeholder="Search gifs..."
                value={ inputValue }
                className={ inputStyle }
                onChange={ handleChange }
            />
        </form>
    );
}

// PROPIEDADES DEL COMPONENTE
/* Tipos de Datos */
AddCategory.propTypes = {
    handleCategory: PropTypes.func.isRequired,
}
