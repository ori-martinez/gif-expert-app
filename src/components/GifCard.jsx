/* Tipos de Datos de las Propiedades del Componente */
import PropTypes from 'prop-types';

// ESTILOS
/* Contenedores */
const gifCardStyle = "bg-violet-500 border-b-4 border-r-4 border-violet-500 rounded-lg shadow-2xl shadow-stone-500";
const gifCardTitleStyle = "h-12 w-full flex items-center justify-center bg-teal-200 border border-t-0 border-stone-900 rounded-b-md";
/* Gifs */
const gifImgStyle = "h-56 w-full border border-b-0 border-stone-900 rounded-t-md object-cover";
/* Texto */
const gifTitleStyle = "text-center text-md text-stone-900 font-semibold";

// TAREAS
/* __________________________________________________________________________________________________________
    
    1. Aladir PropTypes para tener como obligatorias las propiedades title y url
    2. Evaluar el snapshot mediante una prueba
__________________________________________________________________________________________________________ */

// COMPONENTE
/* Carta del Gif */
export const GifCard = ({ title, url }) => {
    // RETORNO
    return (
        <div className={ gifCardStyle }>
            { /* Gif */ }
            <img
                src={ url } 
                alt={ (title === '') ? 'GIF' : title }
                className={ gifImgStyle }
            />

            { /* TÃ­tulo del Gif */ }
            <div className={ gifCardTitleStyle }>
                <span className={ gifTitleStyle }> { (title === '') ? 'GIF' : title } </span>
            </div>
        </div>
    );
}

// PROPIEDADES DEL COMPONENTE
/* Tipos de Datos */
GifCard.propTypes = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
}
