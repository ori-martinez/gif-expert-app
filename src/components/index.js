/* Exportaci√≥n Conjunta de los Componentes */
export * from './AddCategory';
export * from './GifCard';
export * from './GifsGrid';
