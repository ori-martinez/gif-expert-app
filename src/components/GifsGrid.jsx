/* Componentes */
import { GifCard } from './GifCard';
/* Hooks */
import { useFetchGifs } from '../hooks/useFetchGifs';
/* Tipos de Datos de las Propiedades del Componente */
import PropTypes from 'prop-types';

// ESTILOS
/* Contenedores */
const gifTitleStyle = "mx-auto mt-6 w-4/5";
const gifGridStyle = "mx-auto my-4 w-4/5 grid grid-cols-3 gap-12 rounded-lg";
/* Textos */
const h3Style = "ml-12 p-3 text-3xl text-teal-500 font-bold";
const h2Style = "text-center text-4xl text-stone-900 font-bold";

// COMPONENTE
/* Grid de Gifs sobre la Categoría */
export const GifsGrid = ({ gifsCategory, gifsQty }) => {
    /* HOOKS */
    /* Obtención de los Gifs sobre la Categoría */
    const { gifs, isLoading } = useFetchGifs(gifsCategory, gifsQty);

    // RETORNO
    return (
        <>
            { /* Título de la Categoría */ }
            <div className={ gifTitleStyle }>
                <h3 className={ h3Style }> { gifsCategory } </h3>
            </div>

            { /* Mensaje de Carga del Grid de Gifs */
                isLoading && (<h2 className={ h2Style }> Cargando... </h2>)
            }

            { /* Grid de Gifs de la Categoría */ }
            <div className={ gifGridStyle }>
                { gifs.map(gif => (
                    <GifCard
                        key={ gif.id }
                        { ...gif }
                    />
                ))}
            </div>
        </>
    );
}

// PROPIEDADES DEL COMPONENTE
/* Tipos de Datos */
GifsGrid.propTypes = {
    gifsCategory: PropTypes.string.isRequired,
    gifsQty: PropTypes.number.isRequired,
}
