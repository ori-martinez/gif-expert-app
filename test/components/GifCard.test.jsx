import { render, screen } from '@testing-library/react';
/* Components */
import { GifCard } from '../../src/components/GifCard';

describe('Pruebas de <GifCard />', () => { 
    // CONSTANTES
    /* Propiedades del Componente */
    const title = "fullmetal alchemist GIF by Funimation";
    const url = "https://media2.giphy.com/media/BSxFhxneZPCvK/giphy.gif?cid=32e67b780vn6ocq2w42enft34s4oawzg4eyfjpheg4n0mzr2&rid=giphy.gif&ct=g";

    // PRUEBA DE COINCIDENCIA
    test('Debe hacer match con el snapshot', () => { 
        /* Renderización del Componente */
        const { container } = render(<GifCard title={ title } url={ url } />);

        expect(container).toMatchSnapshot();
    });

    // PRUEBA DE COINCIDENCIA DE ATRIBUTOS
    test('La imagen debe ser mostrada con los atributos url y alt indicados', () => {
        /* Renderización del Componente */
        render(<GifCard title={ title } url={ url } />);

        // CONSTANTES
        /* Atributos del Elemento img */
        const { src, alt } = screen.getByRole('img');

        expect(src).toBe(url);
        expect(alt).toBe(title);
    });

    // PRUEBA DE COINCIDENCIA DE TEXTO
    test('El título debe ser mostrado en el componente', () => {
        /* Renderización del Componente */
        render(<GifCard title={ title } url={ url } />);

        
        expect(screen.getByText(title)).toBeTruthy();
    });
})
