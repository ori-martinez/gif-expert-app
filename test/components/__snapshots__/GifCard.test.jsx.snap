// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de <GifCard /> Debe hacer match con el snapshot 1`] = `
<div>
  <div
    class="bg-violet-500 border-b-4 border-r-4 border-violet-500 rounded-lg shadow-2xl shadow-stone-500"
  >
    <img
      alt="fullmetal alchemist GIF by Funimation"
      class="h-56 w-full border border-b-0 border-stone-900 rounded-t-md object-cover"
      src="https://media2.giphy.com/media/BSxFhxneZPCvK/giphy.gif?cid=32e67b780vn6ocq2w42enft34s4oawzg4eyfjpheg4n0mzr2&rid=giphy.gif&ct=g"
    />
    <div
      class="h-12 w-full flex items-center justify-center bg-teal-200 border border-t-0 border-stone-900 rounded-b-md"
    >
      <span
        class="text-center text-md text-stone-900 font-semibold"
      >
         
        fullmetal alchemist GIF by Funimation
         
      </span>
    </div>
  </div>
</div>
`;
